### PR TITLE
Upgrade to GDAL 2.4.1 and introduce Docker image build

### PR DESCRIPTION
**Pull request by jpolchlo**
GDAL 2.3 lacks certain functionality that is needed for proper use of some geotrellis-contrib features.  This PR upgrades to version 2.4.1.

I'm also introducing here a Docker image that depends on the RPM images.

### CODE REVIEW

**Comment by CloudNiner** [comment]
Just noting that this branch built successfully for me and the artifacts are published to the location specified in the new README changes in #4

File: docker/Makefile
Comment (CloudNiner) [inline docker/Makefile:1]: > I'm also introducing here a Docker image that depends on the RPM images.

I'm not totally sure what these artifacts in the `docker` directory are for? It doesn't look like there's any tooling to push or use them anywhere. Perhaps a brief README would be good to have here to clarify how to build them and what the use cases are for them?

File: docker/etc/pam.d/login
Comment (CloudNiner) [inline docker/etc/pam.d/login]: What's this file for? It doesn't look like its mounted or copied into any of the docker containers we build in this PR.

File: docker/Makefile
Comment (CloudNiner) [inline docker/Makefile]: This ran successfully after building the RPMs over in the `rpm` folder.

**Review by CloudNiner — CHANGES_REQUESTED** [review]
RPMs built (although it took twoish hours) and then the image in `./docker` also built successfully.

Most of the comments are just questions/clarifications, but it would be good to get some kind of documentation in the `./docker` folder describing why they exist + what I'd use them for.

File: docker/Makefile
Comment (jpolchlo) [reply docker/Makefile:1]: This is the deployments project, so I figure we're providing the means for future users to deploy to GT-enabled environments, whatever they may be.  We're providing the mattress upon which a future user may dream sweet geodreams with GT.  It's true that this base isn't getting published ATM, and maybe we can fix that part through the CI?  That's something I haven't dealt with yet, so it's not done here since it wasn't priority.

File: docker/etc/pam.d/login
Comment (jpolchlo) [reply docker/etc/pam.d/login]: Sure.  Good question.  This originated from https://github.com/geodocker/geodocker-jupyter-geopyspark, and I think this had to do with the jupyter component of that deployment.  We've moved away from python generally, and therefore, jupyter isn't on the menu either.  Thus, this is vestigial.

File: docker/Makefile
Comment (CloudNiner) [reply docker/Makefile:1]: Ok. Then let's document the state of these images with a README in this folder, and create a new issue that describes what would be done next.  Alternatively, if there's no projects utilizing these images and no ideas or plan for what might use them, then we should just remove them for now, perhaps after archiving them on some other branch in this repo. No reason to clutter up the main branch with code/configuration that isn't in use anywhere.

Comment (jpolchlo) [reply docker/Makefile:1]: Documentation is a good idea, but I disagree with the idea that a general resource like the docker image needs to be justified by projects.  I'd say that this repo is all about making it easy to access a proper environment for developing/deploying GT-based software.  This docker image lowers the barrier to entry, and provides a useful base for development.  It's as justifiable as the `emr` sub-project, and the argument that the few files involved in providing this feature "clutter up" the repo is unconvincing (though I do need to kill that pam login file to honestly make that claim!).  

As far as I can tell, this contribution is the cleanest way we have of providing a good base docker image for GT development (especially after adding GDAL readers to GT), and that's a thing worth keeping around.  And not in some half-supported fashion as it would be if it were relegated to some branch where it would almost surely fall out of sync.

Comment (CloudNiner) [reply docker/Makefile:1]: I'm trying to get at two things:
1. _Why_ do the artifacts in this `./docker` folder exist?
2. If they are useful, _what_ are the next steps for these artifacts that can become follow up issues so that these artifacts can continue to be useful.

I think you've made a case for (1) so that means we're on to (2). So, I still think we should add a README that has a sentence or two describing what you might use this for and how to build it, e.g. "first go to `./rpms` and follow instructions there, then come back here and run `make image`." because otherwise, if the goal is to

> [make] it easy to access a proper environment for developing/deploying GT-based software

then we're certainly not achieving that goal by just having some files in a subfolder that aren't documented in any way -- no one is ever going to find them. 

As a bit of a side question, how does this differ from https://github.com/azavea/docker-openjdk-gdal which is now in use by at least Raster Foundry in production and geotrellis-contrib?

Comment (jpolchlo) [reply docker/Makefile:1]: For sure.  And as I said above, documentation is called for, absolutely.

How this differs from the azavea image (which I had not seen, so thanks for the pointer) is that we provide a more comprehensive loadout of binary dependencies, including PDAL, NetCDF, HDF5, and also spark.  These are all the parts needed to do all the things with GT, I think (with the exception of sonar point clouds via PDAL, as that requires an installation process that I wasn't able to master).

The question of discoverability is a bugaboo, without a doubt.  We've struggled with that in the past in different projects in different ways, and I'm not sure if I were looking for an image to run a GT environment that I would know to look for the image that you cited either.  Seems like the most probable path to finding an image would be stopping on gitter and asking?  One of these images provides a lightweight path to a GDAL-only environment, the other is a bit more heavyweight, but does more.  Horses for courses, I suppose.

The upside of this repo would be that you can test and run in an identical environment to the EMR version (this docker image is based on amazonlinux) without having to spin up EMR, as it's based on exactly the same artifacts.

Comment (CloudNiner) [reply docker/Makefile:1]: Cool -- sounds like we're in agreement. So this PR is waiting on for approval is a short README that describes how to build this container and what it's for in addition to deleting that obsolete file. If that seems like too much of an ask or you think should be a separate issue, lets discuss offline.

> How this differs from the azavea image

👍 makes sense. 

> The question of discoverability is a bugaboo, without a doubt.

A link to this container image and the Azavea one might make sense to add to the locationtech/geotrellis README or documentation somewhere. Otherwise, as you say, yeah it would be very difficult to find them, and you'd have to know where to ask to do so. 

> The upside of this repo would be that you can test and run in an identical environment 

Makes sense and useful to know. Perhaps worth mentioning in the "how to use" section of the aforementioned new README?

Comment (CloudNiner) [reply docker/Makefile:1]: (Created new issue over in GeoTrellis land regarding linking to these two artifacts: https://github.com/locationtech/geotrellis/issues/3099)

**Review by CloudNiner — APPROVED** [review]
🔥 docs notes are succinct but complete.

**Comment by CloudNiner** [comment]
Note to consider rebasing down before merge to keep the commit history clean.

**Comment by jpolchlo** [comment]
Meh.  I'm not sure I see a strong upside for a tidy history on a project such as this.

**Pull request merged**
